test/e2e/es/version_upgrade_test.go

```diff
@@ -0,0 +1,47 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package es
+
+import (
+	"testing"
+
+	"github.com/elastic/cloud-on-k8s/test/e2e/test/elasticsearch"
+)
+
+func TestVersionUpgrade680To720(t *testing.T) {
+	initial := elasticsearch.NewBuilder("test-version-up-3-680-to-720").
+		WithVersion("6.8.0").
+		WithESMasterDataNodes(3, elasticsearch.DefaultResources)
+
+	mutated := initial.WithNoESTopology().
+		WithVersion("7.2.0").
+		WithESMasterDataNodes(3, elasticsearch.DefaultResources)
+
+	RunESMutation(t, initial, mutated)
+}
+
+func TestVersionUpgradeSingle680To720(t *testing.T) {
+	initial := elasticsearch.NewBuilder("test-version-up-1-680-to-720").
+		WithVersion("6.8.0").
+		WithESMasterDataNodes(1, elasticsearch.DefaultResources)
+
+	mutated := initial.WithNoESTopology().
+		WithVersion("7.2.0").
+		WithESMasterDataNodes(1, elasticsearch.DefaultResources)
+
+	RunESMutation(t, initial, mutated)
+}
+
+func TestVersionUpgradeSingle710To720(t *testing.T) {
+	initial := elasticsearch.NewBuilder("test-version-up-1-710-to-720").
+		WithVersion("7.1.0").
+		WithESMasterDataNodes(1, elasticsearch.DefaultResources)
+
+	mutated := initial.WithNoESTopology().
+		WithVersion("7.2.0").
+		WithESMasterDataNodes(1, elasticsearch.DefaultResources)
+
+	RunESMutation(t, initial, mutated)
+}
```

Comment:
Nit: we could try to optimize a bit the number of images used in the e2e tests. By using 7.3.0 instead of 7.2.0 will avoid to pull another version of the ES image. A downside is that it works until the Elastic stack version is 7.3.0. So we could use `test.Ctx().ElasticStackVersion` that should work until we are <=7.x. So maybe we're probably not at a few image pulls near.

Reply:
Yes, these version upgrade tests need a little bit of re-think I believe and not only because of the images we pull. They are going against the grain of the whole test framework by hard coding the version in the test. That means they run redundantly n-times for each stack version we test and so forth.